src/irmin/tree.ml

```diff
@@ -1257,116 +1290,70 @@ module Make (P : Private.S) = struct
-                    k ()
-                | Map children ->
-                    let l = StepMap.bindings children |> List.map snd in
-                    add_steps_to_todo l n k
-                | Value (_, _, Some children) ->
-                    let l =
-                      StepMap.bindings children
-                      |> List.filter_map (function
-                           | _, Node.Remove -> None
-                           | _, Node.Add v -> Some v)
-                    in
-                    add_steps_to_todo l n k)))
-    and add_steps_to_todo : type a. _ -> _ -> (unit -> a Lwt.t) -> a Lwt.t =
-     fun l n k ->
-      (* 1. convert partial values to total values *)
-      let* () =
-        match n.Node.v with
-        | Value (_, _, Some _) ->
-            let+ v = Node.to_value n >|= get_ok "export" in
-            n.v <- Value (repo, v, None)
-        | _ -> Lwt.return_unit
-      in
-      (* 2. push the current node job on the stack. *)
-      let () =
-        match (n.v, Node.cached_value n) with
-        | _, Some v -> Stack.push (add_node n v) todo
-        | Map x, None -> Stack.push (add_node_map n x) todo
-        | _ -> assert false
-      in
-      let contents = ref [] in
-      let nodes = ref [] in
-      List.iter
-        (function
-          | `Contents c -> contents := c :: !contents
-          | `Node n -> nodes := n :: !nodes)
-        l;
-
-      (* 2. push the contents job on the stack. *)
-      List.iter
-        (fun (c, _) ->
-          let h = Contents.hash c in
-          if Hashes.mem seen h then ()
-          else (
-            Hashes.add seen h ();
-            match c.Contents.v with
-            | Contents.Hash _ -> ()
-            | Contents.Value x -> Stack.push (add_contents c x) todo))
-        !contents;
-
-      (* 3. push the children jobs on the stack. *)
-      List.iter
-        (fun n ->
-          Stack.push (fun () -> (add_to_todo [@tailcall]) n Lwt.return) todo)
-        !nodes;
-      k ()
-    in
-
-    let rec loop () =
-      let task = try Some (Stack.pop todo) with Stack.Empty -> None in
-      match task with None -> Lwt.return_unit | Some t -> t () >>= loop
+    let rec on_node (`Node n) k =
+      match n.Node.v with
+      | Node.Hash (_, h) ->
+          (* Not modified *)
+          Node.export ?clear repo n h;
+          k ()
+      | Node.Value (_, v, None) ->
+          (* Not modified *)
+          let h = P.Node.Key.hash v in
+          Node.export ?clear repo n h;
+          k ()
+      | _ ->
+          let* children_seq =
+            let+ seq = Node.seq_unordered n in
+            Seq.map (fun (_, x) -> x) (get_ok "export" seq)
+          in
+          on_node_seq children_seq @@ fun () ->
+          let* v = Node.to_value n in
+          let v = get_ok "export" v in
+          let key = Node.hash n in
+          cnt.node_mem <- cnt.node_mem + 1;
+          let* mem = P.Node.mem node_t key in
```

Comment:
the main reason we check if that node already exists is to avoid iterating on the children if that's the case (e.g. to get an early cut-off)